crates/lib/src/qpu/experimental/random.rs

```diff
@@ -0,0 +1,272 @@
+        let bit = u64::from((seed & base) != 0);
+        acc.bitxor(bit)
+    });
+    (seed << 1) & MAX_SEQUENCER_VALUE | feedback_value
+}
+
+/// This represents the LFSR currently implemented on Rigetti control systems. Specifically,
+/// it implements a 48-bit LFSR with taps at indices 48, 47, 21, and 20.
+#[must_use]
+pub fn lfsr_v1_next(seed: PrngSeedValue) -> u64 {
+    lfsr_next(seed.as_u64, &[48, 47, 21, 20])
+}
+
+fn generate_lfsr_v1_sequence(seed: u64, start_index: u32, series_length: u32) -> Vec<u64> {
+    let mut lfsr = seed & MAX_SEQUENCER_VALUE;
+
+    let range = start_index..(start_index + series_length);
+    let mut collection = vec![];
+    for i in 0..(start_index + series_length) {
+        lfsr = lfsr_next(lfsr, &[48, 47, 21, 20]);
+        if range.contains(&i) {
+            collection.push(lfsr);
+        }
+    }
+    collection
+}
+
+fn prng_value_to_sub_region_index(value: u64, sub_region_count: u8) -> u8 {
+    ((value & MAX_UNSIGNED_MULTIPLIER) % u64::from(sub_region_count))
+        .to_u8()
+        .expect("modulo u8 should always produce a valid value")
+}
+
+/// Given a seed, start index, series length, and sub-region count, this function
+/// will generate the sequence of pseudo-randomly chosen indices Rigetti control
+/// systems.
+///
+/// For instance, if the following Quil program is run for 100 shots:
+///
+/// ```quil
+/// # presumed sub-region size is 3.
+/// DECLARE destination REAL[6] # prng invocations per shot = (6 / sub_region_size)  = 2
+/// DELCARE source REAL[12]     # implicit sub-region count = (12 / sub_region_size) = 4
+/// DECLARE seed INTEGER[1]
+/// DECLARE ro BIT[1]
+///
+/// DELAY 0 1e-6
+///
+// PRAGMA EXTERN choose_random_real_sub_regions "(destination : mut REAL[], source : REAL[], sub_region_size : INTEGER, seed : mut INTEGER)"
+// CALL choose_random_real_sub_regions destination source 3 seed
+/// ```
+///
+/// with a seed of 639523, you could backout the randomly chosen sub-regions with the following:
+///
+/// ```rust
+/// let seed = 639523;
+/// let start_index = 0;
+/// let prng_invocations_per_shot = 2;
+/// let shot_count = 100;
+/// let series_length = prng_invocations_per_shot * shot_count;
+/// let sub_region_count = 4;
+/// let _random_indices = choose_random_real_sub_region_indices(seed, start_index, series_length, sub_region_count).unwrap();
+/// ```
+#[must_use]
+pub fn choose_random_real_sub_region_indices(
+    seed: PrngSeedValue,
+    start_index: u32,
+    series_length: u32,
+    sub_region_count: u8,
+) -> Vec<u8> {
+    generate_lfsr_v1_sequence(seed.as_u64, start_index, series_length)
+        .iter()
+        .map(|&value| prng_value_to_sub_region_index(value, sub_region_count))
+        .collect()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{collections::HashMap, fs::File};
+
+    fn prng_sequences() -> HashMap<u32, Vec<(u64, u64)>> {
+        serde_json::de::from_reader(File::open("tests/prng_test_cases.json").unwrap()).unwrap()
```

Comment:
Others can weigh in, but I believe we generally have a convention of test data files living outside of `src` and then being referenced as e.g.

```rust
concat!(env!("CARGO_MANIFEST_DIR"), "/data/tests/prng_test_cases.json")
```
```